[FEATURE] 노래관리 기능 구현 완료

donsonioc2010 opened the pull request:
## 📝 PR Summary
노래의 관리와 관련된 기능의 구현

#### 🌲 Working Branch
#18 

#### 🌲 TODOs
- [x] Song관련 Entity추가
- [x] 검색어를 바탕으로 노래를 조회하는 기능
- [x] 노래의 세부 페이지를 조회하는 기능
- [ ] 크롤링, 기존에 이미 등록된 번호 가져올 수 있는 방안 찾기.

### Related Issues
- 테스트를 할 방법이 없어서, 세부적으로 해보지는 못했습니다. 추후 확인해봐야 할 듯 합니당..ㅠ
- 배치, 크롤링의 경우 방법은 알겠는데, 따로 티켓을 뺴서 작업해야 할 것 같습니다. 생각보다 작업공수가 소요될 것 같아요.
- 양이 많아서 정신없이 작업해서 문제사항이 많이 보일 수 있습니다. 많은 질타 부탁드립니다.
- application.yml의 경우 env쓰기로해서 설정 추가했습니다.
- db의 경우, local은 h2실행이기에 무조건 create, db쓰는 경우 local-dev는 update가 되도록 변경하였습니다.


### PS
dev 브랜치와 충돌 해결해놨습니다.

donsonioc2010 committed:
fix : SongEntity에서 가사필드 및 컬럼의 오탈자 수정

donsonioc2010 committed:
feat : 관리자 기능에서 사용될 노래의 추가, 수정 삭제 기능 구현

donsonioc2010 committed:
feat : QueryDSL에서 사용할 Bean 생성 및 EntityManager Javax를 받는게 아닌, Jakarta를 사용하도록 변경.

donsonioc2010 committed:
feat : 노래 관리 - 조회 기능 구현 완료

- 일반 사용자, 관리자용 API구현완료
- Env파일 로드 가능하도록 변경.
- DB의 yml의 경우 H2사용은 무조건 Create, DB사용은 update하도록 수정.

donsonioc2010 committed:
Merge branch 'dev' into feat/#18

donsonioc2010 committed:
feat : 누락한 @Validated 추가

JakePark929 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/song/controller/SongAdminController.java:
```diff
@@ -0,0 +1,76 @@
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/admin/songs")
```
> /api/v1/admin/songs -> /admin-api/v1/songs

api 맵핑을 위와 같이 정해보는 것이 어떨까요?
물론 관리 방식 차이지만 로그인을 위해 open 해두는 주소도
open-api/v1/auth 이런 식으로 되어있고 앞 부분의 구분자에 따라
WebConfigure 에서 필터링 하기 쉬울 것으로 예상됩니다!

최종적으로는

- open-api : 모든 사용자에 공개
- api : 로그인 한 사람에만 공개
- admin-api : 관리자 로그인 유저에게만 공개
 
이렇게 필터링 되겠네요!

당장 중요한 부분은 아니니 추후 생각해보도록 합시다 :)

donsonioc2010 replied:
변경완료!

JakePark929 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/song/service/SongManageService.java:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * 요청자가 관리지인지, 일반사용자인지 여부를 확인한다.
+     */
+    private void isValidRequestUserAdmin(UserDto user) {
```
이 부분 #24 즐겨 찾기 PR 에서 상태 체크를 로직 에서 처리할 지 의견 주셨던 부분과 동일한 의견인데요.

저도 어제 이 부분 보면서 사용자 권한은 필터에서 처리하는 것이 맞겠다고 생각했었습니다.
같은 내용으로 필터에서 상태도 처리할 수 있을 것으로 보입니다!

권한 확인을 매 서비스 로직에서 수행하면 어떤 문제가 있을지는 
의견을 주셔서 잘 아실 것이라고 생각합니다 :)

donsonioc2010 replied:
해당 로직은 제거 완료해놨습니다.

@AdminUser 이런 어노테이션을 제작할까 했는데, 관리자 기능 PR추가한거 보고 작업은 따로 하지는 않았습니다 ㅎㅎ :)

JakePark929 replied:
저도 어노테이션 방식으로 할까 했는데 굳이 싶어서 인터셉터로 처리했습니다.
서비스가 확장된다면 어노테이션 방식도 필요해질 수도 있겠네요!

JakePark929 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/song/repository/SongCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,83 @@
+import org.springframework.util.ObjectUtils;
+
+@RequiredArgsConstructor
+public class SongCustomRepositoryImpl implements SongCustomRepository{
```
QueryDSL 에 페이지까지 녹여서 잘 써주셨네요!👍

제가 썼던 projection 방식도 있지만, 
결국 한계가 있기 때문에 이런 큰 쿼리는 QueryDSL 쓰는 것이 맞다고 봅니다. 😄

JakePark929 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/song/business/SongBusiness.java:
```diff
@@ -0,0 +1,48 @@
+    private final SongSelectService songSelectService;
+    private final SongManageService songManageService;
+
+    @Transactional(readOnly = true)
```
저는 트랜잭션 처리를 각 서비스 클래스에서 해주었는데요!
추후 비즈니스 로직이 어떻게 될 지 모르겠지만,
비즈니스 로직이 커지면 트랜잭션 관리를 어떻게 해야 될지 생각해봐야 할 것 같습니다.

donsonioc2010 replied:
해당 부분은 어떤 기능을 구현하냐에 따라 달라질 것 같습니다.

비즈니스 전체 롤백이 필요한 경우에는 비즈니스로직이 트랜잭션을 잡아야 할 것 같고, 롤백이 필요없는 경우 각 서비스 클래스가 잡아도 괜찮지 않을까 하네요. :)

근데 생가갷보니, 저도 비즈니스 로직에서 트랜잭션을 잡지 않아도 될 것 같아보이는데, 버릇마냥 다잡아 버렸네요..😂

JakePark929 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/song/converter/SongConverter.java:
```diff
@@ -0,0 +1,90 @@
+@Slf4j
+@Converter
+@RequiredArgsConstructor
+public class SongConverter {
```
저는 컨버터 이름 짓기 귀찮아서 toEntity, toDto 로 오버로딩 시켜 버렸는데요.
이름으로 구분 해주는 것도 아주 좋습니다! 👍👍

JakePark929 reviewed (approved):
확실히 작업량이 많아서 생각보다 한참 봤습니다.
크게 수정할 부분은 없어 어프루브 해 놓겠습니다.
코멘트 확인하시고, 머지 진행하시면 되겠습니다!

발리데이션 어노테이션은 어딨지 했는데 다 쓰고 보니 바로 커밋하셨네요!

고생하셨습니다~ :)

donsonioc2010 committed:
feat : admin-api의 RequestMapping 변경

donsonioc2010 committed:
feat : 관리자 검증 여부로직 제거

donsonioc2010 committed:
feat : PageResponse는 불필요 할 것 같아 제거함.

donsonioc2010 committed:
feat : 설정 변경

- QueryDSLConfig에서 JPQLTemplates.DEFAULT속성을 추가안하는 경우 SpringBoot3에서 QueryDSL의 기존 기능들이 작동하지 않는 이슈로 추가.
- 완성된 Page객체를 다시 Response시 직렬화 이슈가 발생 가능하다는 Spring권고에 따라 Api모듈에 @EnableSpringDataWebSupport 어노테이션 추가.

donsonioc2010 committed:
Chore : 나무위키에서 크롤링해 1차 가공한 노래 데이터 쿼리 추가

- 금영 약 7천곡
- TJ약 7만곡

donsonioc2010 committed:
fix : 기능 이슈 수정

- SongResponse @Getter없어서 Response내용 없었음..ㅋ;;
- new PageImpl사용이 아닌, PageableExcutionUtils를 사용하는 것으로 변경.
- ..or조건 사용해야하는데... and조건 사용해서 결과가 안나오는 이슈 수정...ㅎ;;

donsonioc2010 reviewed (commented):
노래 조회안되던 부분 수정완료해놨습니다.

3시까지 추가적인 코멘트 없으면 Merge하도록 하겠습니당 :)

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/song/business/SongBusiness.java:
```diff
@@ -0,0 +1,48 @@
+
+@Slf4j
+@Business
+@Transactional
```
클래스 레벨에서 @Transactional 을 사용하셨는데, 메서드 레벨에서도 추가로 @Transactional을 사용하신 이유가 있으신가요?

donsonioc2010 replied:
아 해당부분은... 제가 삭제를 잊었습니다. 감사합니당 :)

Business레이어는 메소드별로 트랜잭션을 사용하려 의도해서 메서드레벨에 전부 작성했습니다.

birdieHyun commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/song/repository/SongCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,84 @@
+                    .or(artistContains(keyword))
+            )
+            .orderBy(song.id.asc())
+            .distinct()
+            .offset(pageable.getOffset())
+            .limit(pageable.getPageSize())
+            .transform(
```
jpql 로 페이징 쿼리를 작성하게 되면, native 쿼리로도 offset 과 limit이 사용되는걸로 기억하고 있는데, 
점차 offset이 늘어나게 된다면 단점은 없을까요?

donsonioc2010 replied:
`offset`에 있어서 페이지가 늘어났을떄의 이슈에 대해서는 처음 알게 되었고, 아래의 글들을 참조해 볼 수 있었습니다.

- https://jojoldu.tistory.com/528
- https://jojoldu.tistory.com/529
- https://jojoldu.tistory.com/530
- https://jojoldu.tistory.com/531

멘토님 덕에 어울리는 방법이 어떤게 있을까를 생각해보게 되었고, 일단 Pageable객체를 사용하는 쪽으로 구현을 전부 하고있기에, NoOffset 방식을 고려하긴 힘들거라 생각이 들었고, 인덱스를 염두는 해두고 있었기에 커버링 인덱스쪽으로 작업을 진행하는게 더 도움이 될 거라 판단을 할 수 있었습니다.

그에 맞춰 쿼리를, id를 받아 in절을 직접 조회하는 방식으로 변경하는 작업을 할 수 있었습니다.

JakePark929 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/config/SpringDataWebConfig.java:
이 부분 꽤 중요한 이슈네요!
확인 하시느라 고생하셨습니다!!

donsonioc2010 committed:
fix : Offset관련 이슈 수정

- @Transactional이 SongBusiness의 Class레벨에 작성된 부분 수정.
- SongCustomRepositoryImpl의 Offset부분에 대하여, 조회 쿼리 변경.

JakePark929 reviewed (commented):
👍

donsonioc2010 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/config/QueryDSLConfig.java:
```diff
@@ -0,0 +1,21 @@
+    private EntityManager entityManager;
+
+    @Bean
+    public JPAQueryFactory jpaQueryFactory() {
+        return new JPAQueryFactory(JPQLTemplates.DEFAULT, entityManager);
+    }
```
해당 변경점에 대하여 참고한 레퍼런스 자료 링크는 아래 첨부드립니다.

- https://hojun-dev.tistory.com/entry/JAVA-QueryDsl-transform-%EB%B0%8F-SqmCaseSearched-%EC%98%A4%EB%A5%98-%ED%95%B4%EA%B2%B0%EB%B0%A9%EB%B2%95-with-Hibernate-6x
- https://github.com/querydsl/querydsl/issues/3428

donsonioc2010 commented:
추후 노래관리 쿼리들 분석후 인덱스작업 필요 :)
